Snap 2919 : Implementation of Structured Streaming UI Tab

snappy-sachin opened the pull request:
## What changes were proposed in this pull request?

Implementation of the Structure Streaming UI Tab to let users monitor the structured streaming query/application statistics and progress .
Structured Streaming Tab is available both in SnappyData embedded cluster as well as in smart connector application (using Snappy Spark distribution)

Structured Streaming Tab has below capabilities:
 - Listing all Structured Streaming Queries/Applications submitted to SnappyData cluster using submit-job command. Similarly in smart connector this tab will list streaming queries executed in cluster.
 - Allows user selecting queries from left hand side navigation panel, to view details view on right side main query details panel.
 - Query details panel displays selected queries details and statistics, as listed below;
   -- Query Name if provided, Query Id otherwise
   -- Start Date & Time
   -- Up time
   -- Trigger Interval
   -- Batches Processed
   -- Status
   -- Total Input Records
   -- Current Input Rate
   -- Current Processing Rate
   -- Total Batch Processing Time
   -- Avg. Batch Processing Time
 - Query details panel also lists sources of streaming query along with each source details like type, description, input records, input and processing rate
 - Query details panel also displays sink details of streaming query.
 - Query details panel depicts structured streaming queries behaviourial trends using following
   -- Input Records on every batch
   -- Input Rate vs Processing Rate
   -- Processing Time
   -- Aggregation State, if available

Please check JIRA item SNAP-2919 for UI screenshots (https://jira.snappydata.io/browse/SNAP-2919)

## How was this patch tested?

 - Tested manually

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

snappy-sachin committed:
Code changes for SNAP-2919:
 - Adding streaming page CSS and JavaScript files.

snappy-sachin committed:
Intermediate code changes:
 - Auto-refresh feature.
 - Query stats populating on UI.
 - CSS changes.

snappy-sachin committed:
- CSS changes

snappy-sachin committed:
Code changes for:
 - CSS changes
 - Display Status in different colors
 - Rounding of float values on UI
 - Display numbers with thousands separator mark.
 - Display time unit as ms for milliseconds
 - Display sources table with basic stats.
 - Display sink description
 - Display Aggregation state (state operator) charts.

snappy-sachin committed:
CSS and JavaScript code changes for highlighting query selection in query navigation panel.

snappy-sachin committed:
Code changes for:
 - Adding utility function for conversion of duration time in human readable form.
 - JavaScript changes for displaying queries latest input rate and processing rate.

snappy-sachin committed:
- CSS and JavaScript changes for displaying connecton error message.

snappy-sachin committed:
Code changes for:
 - Display aggregation states chart only when applicable
 - Removed updated records/rows trend line from aggregation states chart
 - Fixing few other issues

snappy-sachin committed:
- JavaScript changes for displaying sink specific details.

snappy-sachin committed:
Code changes for displaying sink details table.

snappy-sachin committed:
Code Changes for SNAP-2919:
 - Moving below classes/files from SnappyData Core to Spark
   SnappyStreamingQueryListener.scala,
   StreamingRepository.scala,
   SnappyStreamingApiRootResource.scala,
   StreamsInfoResource.scala,
   streamapi.scala,
   SnappyStreamingTab.scala,
   SnappyStructuredStreamingPage.scala
 - Moving Code for updating UI for Structured Streaming tab from SnappySession to SparkSession

snappy-sachin committed:
Merge branch 'snappy/branch-2.1' into SNAP-2919

snappy-sachin committed:
Some code cleanup.

snappy-sachin committed:
Merge branch 'snappy/branch-2.1' into SNAP-2919

snappy-sachin committed:
- Code clean up and code refactoring.

snappy-sachin committed:
Code changes:
 - In spark code, updating QueryStartedEvent signature to include triggerInterval
 - Changes for displaying Processing Threshold line in Stream Processing Time chart on UI
 - Adding Trigger Interval on UI

suranjan commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/SnappyStreamingQueryListener.scala:
```diff
@@ -0,0 +1,73 @@
+      val sqs = streamingRepo.allQueries.get(pr.id).get
+      sqs.updateQueryStatistics(event)
+    } else {
+      val queryName = {
```
I see that we are trying to put query stats again for new query id.
Is there a case where onQueryprogress is called without onQueryStarted?

snappy-sachin replied:
Discussed this with Vatsal. 
There is rare possibility of queryStartEvent is not posted or missed but such case is not rquired to be handled. Hence just adding warning log statement for the same. Counter code implantation is removed.

suranjan commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingRepository.scala:
```diff
@@ -0,0 +1,177 @@
+    startTime: Long,
+    trigger: Trigger = ProcessingTime(0L)) {
+
+  private val MAX_SAMPLE_SIZE = 100
```
Should this be configurable?

snappy-sachin replied:
Done as suggested.
Added **_spark.sql.streaming.uiTrendsMaxSampleSize_** configuration parameter to configure how many historic data points be plotted on trends charts on structure streaming UI.
_**Note:**
In my opinion we should keep this parameter for internal use and let user not be aware of this config parameter. This is because Streaming UI charts are designed to fit in very small and limited space on UI page. Therefore large data sample if configured then charts will not be able to conveniently plot them._

suranjan commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingRepository.scala:
```diff
@@ -0,0 +1,177 @@
+  def updateQueryStatistics(event: StreamingQueryListener.QueryProgressEvent): Unit = {
+    val progress = event.progress
+
+    if(this.currentBatchId < progress.batchId){
```
Please see if we need to ignore or overwrite the existing stats in case of event for older batch id.

snappy-sachin replied:
Although I have implemented code changes to handle restart query case, I haven't added those changes yet. Some more discussion is needed.
So I suggest not to include this in current release. We can take it as an enhancement in next release.

suranjan replied:
Ok. Lets file that as a separate task and checkin.

suranjan commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/SnappyStreamingQueryListener.scala:
```diff
@@ -0,0 +1,73 @@
+      }
+    }
+
+    streamingRepo.allQueries.put(event.id,
```
Dont see where we remove from allQueries, this could be memory leak in long running cluster?

snappy-sachin replied:
Done.
Also added spark configuration paramenter **_spark.sql.streaming.uiRunningQueriesDisplayLimit_** to configure how many maximum number of running queries be displayed on structure streaming UI.

suranjan commented on sql/core/src/main/scala/org/apache/spark/status/api/v1/StreamsInfoResource.scala:
```diff
@@ -0,0 +1,39 @@
+import org.apache.spark.sql.streaming.StreamingRepository
+
+@Produces(Array(MediaType.APPLICATION_JSON))
+private[v1] class StreamsInfoResource {
```
Please add test for this api? Can be added as an example for user.

snappy-sachin replied:
Yes, planning to add it after the merge and documentation.

vatsalmevada commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala:
```diff
@@ -96,7 +96,8 @@ object StreamingQueryListener {
       val runId: UUID,
-      val name: String) extends Event
+      val name: String,
+      val trigger: Trigger = ProcessingTime(0L)) extends Event
```
Do we need this default value? Is it set to avoid compilation failures at other usages(scala tests)?

snappy-sachin replied:
Yes. That is true.

snappy-sachin committed:
Code changes for SNAP-2919 :
  - Feedback Question "Is there a case where onQueryprogress is called without onQueryStarted?"
    Ans: As per discussion, YES it can happen in rare situation but no need to handle such case.
      So removing the counter code implementation (on queryProgressEvent) of handling missed queryStartEvent.
      Instead, now just logging warning message for the same.
  - Adding two configurable parameters in sparks SQLConf.scala
      1) spark.sql.streaming.uiRunningQueriesDisplayLimit :
           To configure how many queries be displayed on structure streaming UI.
      2) spark.sql.streaming.uiTrendsMaxSampleSize :
           To configure how many historic data points be plotted on trends charts on structure streaming UI.
  - Handling of query removal in case of uiRunningQueriesDisplayLimit limit is reached.
      Inactive queries are removed if there is no space for newly added running query.
      If all existing queries are active and uiRunningQueriesDisplayLimit limit is reached then
      newly added query won't be displayed on UI.
  - Fixed issue - Query details panel keeps displaying old inactive query details if that query was
      selected before it was removed from query navigation panel.
  - Code refactoring
